File: Pythoncode/64.最小路径和.py
class Solution:
    def minPathSum(self, grid):
        d = {}
        return self.helper(grid, len(grid)-1, len(grid[0])-1, d)
    def helper(self, grid, i, j, d):
        if (i, j) in d:
            return d[(i, j)]
        if i < 0 or j < 0:
            return None
        left = self.helper(grid, i-1, j, d)
        up = self.helper(grid, i, j-1, d)
        if left == None and up == None:
            d[(i, j)] = grid[i][j]
        elif left == None and up != None:
            d[(i, j)] = grid[i][j] + up
        elif up == None and left != None:
             d[(i, j)] = grid[i][j] + left
        else:
             d[(i, j)] = grid[i][j] + min(left, up)
        return d[(i, j)]
s = Solution()
s.minPathSum([[1,3,1],[1,5,1],[4,2,1]])

File: Cppcode/64.最小路径和.cpp
#include <climits>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        vector<vector<int>> pathSum(grid.size(), vector<int>(grid[0].size()));
        for (auto i = 0; i < grid.size(); i++)
        {
            for (auto j = 0; j < grid[0].size(); j++)
            {
                if (i == 0 && j == 0)
                {
                    pathSum[i][j] = grid[i][j];
                    continue;
                }
                auto a = i-1>=0 ? pathSum[i-1][j] : INT_MAX;
                auto b = j-1>=0 ? pathSum[i][j-1] : INT_MAX;
                pathSum[i][j] = min(a, b) + grid[i][j];
            }
        }
        return pathSum[grid.size()-1][grid[0].size()-1];
    }
};